server/src/infra/repositories/database.repository.ts

```diff
@@ -210,6 +210,25 @@ export class DatabaseRepository implements IDatabaseRepository {
     return res as R;
   }
 
+  async withTryLock<R>(lock: DatabaseLock, callback: () => Promise<R>): Promise<R> {
+    let res;
+    const queryRunner = this.dataSource.createQueryRunner();
+    try {
+      const lockAcquired = await this.tryLock(lock, queryRunner);
+      if (lockAcquired) {
+        res = await callback();
+      }
+    } finally {
+      try {
+        await this.releaseLock(lock, queryRunner);
+      } finally {
+        await queryRunner.release();
+      }
+    }
+
+    return res as R;
```

Comment:
This is not actually true in the case the lock was not acquired. Maybe we should change the return type to be something like this instead:

```ts
// : Promise<TypeLockResponse<R>> { ...
type TryLockResponse<R> = { response: R | null; acquired: boolean }
```

Reply:
Check the latest version, I already changed this